### PR TITLE
feat: NGSCheckMate make pattern

### PR DESCRIPTION
**Pull request by tdayris**
Adds the companion script [maktsvnpattern](https://github.com/parklab/NGSCheckMate/blob/master/patterngenerator/makesnvpattern.pl) for fastq analysis with NGSCheckMate.

### QC

* [X] I confirm that I have followed the [documentation for contributing to `snakemake-wrappers`](https://snakemake-wrappers.readthedocs.io/en/stable/contributing.html).

While the contributions guidelines are more extensive, please particularly ensure that:
* [X] `test.py` was updated to call any added or updated example rules in a `Snakefile`
* [X] `input:` and `output:` file paths in the rules can be chosen arbitrarily
* [X] wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`)
* [X] temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to 
* [X] the `meta.yaml` contains a link to the documentation of the respective tool or command under `url:`
* [X] conda environments use a minimal amount of channels and packages, in recommended ordering

## Summary by CodeRabbit

- **New Features**
	- Introduced the `NGSCheckMate MakeSNVPattern` tool for generating SNP pattern files.
	- Added a new `environment.yaml` file to streamline package management.
	- Created a wrapper for the `makesnvpattern.pl` script to facilitate output generation.
	- Added input and output parameters for the SNP pattern generation process.
	- Introduced a new file for creating a specific Conda environment for Linux 64-bit systems.
  
- **Bug Fixes**
	- Enhanced logging and error handling during output file generation.

- **Tests**
	- Added a new test function to improve testing coverage for the SNP pattern generation functionality.

- **Data Updates**
	- Expanded genomic variant dataset with new entries in the `variants.bed` file.
	- Added reference files, including `genome.fasta` and `genome_bwt.4.ebwt`, for testing purposes.

### CODE REVIEW

File: bio/ngscheckmate/makesnvpattern/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+index = commonprefix(snakemake.input.bowtie).rstrip(".")
+
+with TemporaryDirectory() as tempdir:
+    shell(
+        "makesnvpattern.pl "
+        "{snakemake.input.bed} "
+        "{snakemake.input.fasta} "
+        "{index} {tempdir} snake_out {log}"
+    )
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngscheckmate/makesnvpattern/wrapper.py:24]: Input file validation is made by Snakemake.

File: bio/ngscheckmate/makesnvpattern/test/Snakefile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        ),
+    output:
+        pattern="genome.pt",
+        pattern_text=temp("genome.txt"),
+        fasta=temp("genome.pt.fasta"),
+        bowtie=temp("genome.pt.bowtie"),
+        ntm=temp("genome.pt.ntm"),
+        uniq=temp("genome.pt.ntm.uniq"),
+        txt_sorted=temp("genome.txt.sorted"),
+        pattern_sorted=temp("genome.pt.txt.sorted"),
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngscheckmate/makesnvpattern/test/Snakefile:22]: These are examples, IO can be chosen freely

File: bio/ngscheckmate/makesnvpattern/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    if text:
+        shell("mv --verbose '{tempdir}/snake_out.txt' {text} {log}")
+
+    uniq = snakemake.output.get("uniq")
+    if uniq:
+        shell("mv --verbose '{tempdir}/snake_out.uniq.txt' {uniq} {log}")
+
+    txt_sorted = snakemake.output.get("txt_sorted")
+    if txt_sorted:
+        shell("mv --verbose '{tempdir}/snake_out.uniq.txt.sorted' {txt_sorted} {log}")
```
[the remark this answers is not in the document]
Comment (tdayris) [reply bio/ngscheckmate/makesnvpattern/wrapper.py]: Done.

**Comment by fgvieira** [comment]
Can you pin the environment?

File: bio/ngscheckmate/makesnvpattern/wrapper.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+index = commonprefix(snakemake.input.bowtie).rstrip(".")
+
+with TemporaryDirectory() as tempdir:
+    shell(
+        "makesnvpattern.pl "
+        "{snakemake.input.bed} "
+        "{snakemake.input.fasta} "
+        "{index} {tempdir} snake_out {log}"
+    )
```
[the remark this answers is not in the document]
Comment (fgvieira) [reply bio/ngscheckmate/makesnvpattern/wrapper.py:23]: This wrapper has no extra parameters.

File: bio/ngscheckmate/makesnvpattern/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+__author__ = "Thibault Dayris"
+__copyright__ = "Copyright 2024, Thibault Dayris"
+__email__ = "thibault.dayris@gustaveroussy.fr"
+__license__ = "MIT"
+
+from snakemake.shell import shell
+from tempfile import TemporaryDirectory
+from os.path import commonprefix
+
+# Optional parameters
```
Comment (fgvieira) [inline bio/ngscheckmate/makesnvpattern/wrapper.py:12]:
```suggestion
```

File: bio/ngscheckmate/makesnvpattern/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,21 @@
+name: NGSCheckMate MakeSNVPattern
+url: https://github.com/parklab/NGSCheckMate?tab=readme-ov-file#1-patterngenerator
+description: Generate SNP pattern file
+authors:
+  - Thibault Dayris
+input:
+  - bed: Path to bed intervals
+  - fasta: Path to fasta genome sequence
+  - bowtie: List of paths to bowtie index files
```
Comment (fgvieira) [inline bio/ngscheckmate/makesnvpattern/meta.yaml:9]: Maybe rename it to `index`?

File: bio/ngscheckmate/makesnvpattern/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+output:
+  - pattern: Path to binary pattern file. Main output file.
+  - pattern_text: Path to genomic regions extracted from bed intervals
+  - fasta: Path to fasta-formatted regions extracted from bed intervals
+  - bowtieout: Path to bowtie aligned extracted regions
+  - ntm: Path to parsed bowtie output
+  - uniq: Path to filtered bowtie output (used for internal patterns checks only)
+  - txt_sorted: Path to sorted filtered bowtie output (used for internal patterns checks only)
+  - pattern_text: Path to extracted patterns (used for internal patterns checks only)
```
Comment (fgvieira) [inline bio/ngscheckmate/makesnvpattern/meta.yaml:18]: Duplicated.
```suggestion
```

Comment (fgvieira) [inline bio/ngscheckmate/makesnvpattern/meta.yaml:18]: Could you try improving the description a bit? Right now it is not clear how many types of outputs there are. From the description"
- Are the same as regions? That is, do`pattern` and `pattern_text` have the same info just in diff formats? If so, `pattern` (raw binary), `pattern_text` (text? what format?), `pattern_sort` (text? sorted what format?), `fasta` (text fasta format)
- bowtie2 output: `bowtieout` (raw), `ntm` (parsed), `uniq` (uniq but parsed?), `txt_sorted` (filtered and sorted, but parsed?)

**Review by fgvieira — APPROVED** [review]

File: bio/ngscheckmate/makesnvpattern/meta.yaml (hunk 7 above)
Comment (fgvieira) [reply bio/ngscheckmate/makesnvpattern/meta.yaml]: That looks perfect. Thanks for your work!

**Pull request merged**